File: app/src/main/java/org/androidbootmanager/app/ConfigFile.java
package org.androidbootmanager.app;

import android.annotation.SuppressLint;

import com.topjohnwu.superuser.io.SuFile;
import com.topjohnwu.superuser.io.SuFileInputStream;

import java.io.FileNotFoundException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;

@SuppressWarnings("ResultOfMethodCallIgnored")
public class ConfigFile {
    Map<String, String> data = new HashMap<>();

    String get(String name) {
        return data.get(name);
    }

    void set(String name, String value) {
        data.put(name, value);
    }

    String exportToString() {
        StringBuilder out = new StringBuilder();
        for (String key : data.keySet()) {
            out.append(key).append(" ").append(get(key)).append("\n");
        }
        return out.toString();
    }

    void exportToFile(File file) {
        try {
            if (!file.exists()) file.createNewFile();
            PrintWriter out = new PrintWriter(file);
            out.write(exportToString());
            out.flush();
            out.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @SuppressLint("SdCardPath")
    void exportToPrivFile(String tempfilename, String fullPath) {
        exportToFile(new File("/data/data/org.androidbootmanager.app/files/" + tempfilename));
        SuFile.open("/data/data/org.androidbootmanager.app/files/" + tempfilename).renameTo(new File(fullPath));
    }

    static ConfigFile importFromString(String s) {
        ConfigFile out = new ConfigFile();
        for (String line : s.split("\n")) {
            line = line.trim();
            out.set(line.substring(0, line.indexOf(" ")), line.substring(line.indexOf(" ")).trim());
        }
        return out;
    }

    static ConfigFile importFromFile(File f) throws ActionAbortedCleanlyError {
        StringBuilder s = new StringBuilder();
        SuFileInputStream i;
        byte[] b = new byte[1024];
        try {
            i = new SuFileInputStream(f);
        } catch (FileNotFoundException e) {
            throw new ActionAbortedCleanlyError(e);
        }
        while (true) {
            try {
                if (!(i.read(b) > 1)) break;
                s.append(Arrays.toString(b));
            } catch (IOException e) {
                throw new ActionAbortedCleanlyError(e);
            }
        }
        return importFromString(s.toString());
    }

    static ConfigFile importFromFile(String s) throws ActionAbortedCleanlyError {
        return importFromFile(new File(s));
    }
}

File: app/src/main/java/org/andbootmgr/app/util/ConfigFile.kt
package org.andbootmgr.app.util

import com.topjohnwu.superuser.io.SuFileInputStream
import com.topjohnwu.superuser.io.SuFileOutputStream
import org.andbootmgr.app.ActionAbortedCleanlyError
import org.andbootmgr.app.ActionAbortedError
import java.io.*

class ConfigFile {
	val data: MutableMap<String, String> = HashMap()
	operator fun get(name: String): String? {
		return data[name]
	}

	operator fun set(name: String, value: String) {
		data[name] = value
	}

	fun exportToString(): String {
		val out = StringBuilder()
		for (key in data.keys) {
			out.append(key).append(" ").append(get(key)).append("\n")
		}
		return out.toString()
	}

	@Throws(ActionAbortedError::class)
	fun exportToFile(file: File) {
		try {
			val outStream = SuFileOutputStream.open(file)
			outStream.write(exportToString().toByteArray())
			outStream.close()
		} catch (e: IOException) {
			throw ActionAbortedError(e)
		}
	}

	fun has(s: String): Boolean {
		return data.containsKey(s)
	}

	companion object {
		fun importFromString(s: String): ConfigFile {
			val out = ConfigFile()
			var line: String
			for (lline in s.split("\n").toTypedArray()) {
				line = lline.trim()
				val delim = line.indexOf(" ")
				if (delim != -1)
					out[line.substring(0, delim)] =
						line.substring(delim).trim()
			}
			return out
		}

		@Throws(ActionAbortedCleanlyError::class)
		fun importFromFile(f: File): ConfigFile {
			val s = ByteArrayOutputStream()
			val b = ByteArray(1024)
			var o: Int
			val i: InputStream = try {
				SuFileInputStream.open(f)
			} catch (e: FileNotFoundException) {
				throw ActionAbortedCleanlyError(e)
			}
			while (true) {
				try {
					if (i.read(b).also { o = it } <= 1) break
					s.write(b, 0, o)
				} catch (e: IOException) {
					throw ActionAbortedCleanlyError(e)
				}
			}
			return importFromString(s.toString())
		}

		@Throws(ActionAbortedCleanlyError::class)
		fun importFromFile(s: String): ConfigFile {
			return importFromFile(File(s))
		}
	}
}
